### PR TITLE
Golang coverage report test

### PR DESCRIPTION
**Pull request by catenacyber**
This is a draft, not to be merged yet, following the discussion in #2817

The result now is that 
`python infra/helper.py build_fuzzers --sanitizer coverage gonids`
builds a coverage report named `build/out/gonids/covfuzzparserule.html`

There is still a big list of points to address :
- We need to have the corpus directory when running `python infra/helper.py build_fuzzers`. For now, the directory is hardcoded to `$OUT/corpus/$fuzzer/` . Is this possible and the right thing to do ?
- There is no need to run `python infra/helper.py coverage gonids` because this relies on go testing which needs the source code. I guess that there is somewhere where we should add a test like if we already have some html report, no need to launch this...
- should ossfuzz_test.go be part of the docker image (and not related to this specific project) ? I think so
(sorry if I do not have the right wording about Docker)
- What about the naming of the file for the coverage report ? Is it ok ?
- The modifications to build.sh are meant to be generic. Should this be left to be customized by individual projects ? Or should there be a helper script `compile_golang` with this `compile_fuzzer` function ? 

I think that the alternative solution to have a coverage report while not using go test, is to modify go-fuzz-build, and that means a bigger work.
Since there is `LLVMFuzzerInitialize` but no such function for closing, the best shot would be to modify go-fuzz `LLVMFuzzerTestOneInput` https://github.com/dvyukov/go-fuzz/blob/master/go-fuzz-build/main.go#L876 so that coverage info gets streamed after processing each input, and I did not find out yet to access this information.
Then a later processing could transform this raw coverage profiles into some nice html report.

PS : There is the bug #3107 which seems to prevent gonids project to have a fresh generated corpus

### CODE REVIEW

**Comment by catenacyber** [comment]
Happy new year and ping @Dor1s

**Comment by Dor1s** [comment]
@catenacyber we can land this, but our infra will not be running it in the meantime anyway. Please let us know if you want this landed

**Comment by catenacyber** [comment]
Thanks, this is only a draft for the moment, not to be landed right away.
I would rather like to know how we can make your infra run this.

Can you answer the questions in my bullets list ? such as about the naming of the file for the coverage report ?
Or am I too far and missing something else ?

**Comment by Dor1s** [comment]
> - We need to have the corpus directory when running `python infra/helper.py build_fuzzers`. For now, the directory is hardcoded to `$OUT/corpus/$fuzzer/` . Is this possible and the right thing to do ?
> - There is no need to run `python infra/helper.py coverage gonids` because this relies on go testing which needs the source code. I guess that there is somewhere where we should add a test like if we already have some html report, no need to launch this...

We would need to copy the Go sources into `$OUT` by extending this command: https://github.com/google/oss-fuzz/blob/96c9cd26c2e77a85f5ade9e50296edd20b64cac6/infra/base-images/base-builder/compile#L80

That would let us to have sources in the base-runner image, and then we would use `coverage` command to actually generate the report, instead of doing that in `build_fuzzers`.

> - should ossfuzz_test.go be part of the docker image (and not related to this specific project) ? I think so
(sorry if I do not have the right wording about Docker)

Yes, it's fine to have it in the image.

> - What about the naming of the file for the coverage report ? Is it ok ?

What name do you mean?

> - The modifications to build.sh are meant to be generic. Should this be left to be customized by individual projects ? Or should there be a helper script `compile_golang` with this `compile_fuzzer` function ? 

Yeah, we definitely have to generalize it somewhere. With #3297 being addressed, it should be fine to have something like `compile_go` that would take care of common Go specifics.

**Comment by catenacyber** [comment]
Thanks for your review, I will rework this PR...

>> What about the naming of the file for the coverage report ? Is it ok ?
> What name do you mean?

I meant the name of the generated html file report :
https://github.com/google/oss-fuzz/pull/3142/files#diff-3116338d1d486a71c32a97ec981199deR32

**Comment by Dor1s** [comment]
The doesn't really matter, as we'll likely just add a dummy `index.html` that would redirect to the actual report, e.g. for C/C++ coverage we have:

```
$ curl https://storage.googleapis.com/oss-fuzz-coverage/libxml2/reports/20200203/linux/index.html

    <!DOCTYPE html>
    <html>
      <head>
        <!-- HTML meta refresh URL redirection -->
        <meta http-equiv="refresh" content="0; url=directory_view_index.html">
      </head>
    </html>
```

Just because the underlying coverage implementation might change, but the `index.html` file should always be present and should send the user to the right page

**Comment by catenacyber** [comment]
With this latest commit, we are getting close I think :

I modified base-runner and base-builder so, I needed to run 
```
docker build -t gcr.io/oss-fuzz-base/base-builder infra/base-images/base-builder
docker build -t gcr.io/oss-fuzz-base/base-runner infra/base-images/base-runner
```
To get going.

After downloading the corpus to build/corpus/gonids/fuzz_parserule/
The documented commands work :
```
PROJECT_NAME=gonids
python infra/helper.py build_fuzzers --sanitizer=coverage $PROJECT_NAME
python infra/helper.py coverage --no-corpus-download $PROJECT_NAME
```
ie, I get the html report named as build/out/gonids/report/index.html

What is left to be done ?

> With #3297 being addressed, it should be fine to have something like compile_go that would take care of common Go specifics.

The `compile_fuzzer` function is only used by this gonids project, and should be used by all golang projects

The big remaining question is I am not sure how to deal with multiple fuzz targets...

What do you think @Dor1s ?

File: infra/base-images/base-builder/Dockerfile
Comment (catenacyber) [inline infra/base-images/base-builder/Dockerfile]: Not sure why I needed this but `docker build -t gcr.io/oss-fuzz-base/base-builder infra/base-images/base-builder` did not work without it

File: infra/base-images/base-runner/Dockerfile
Comment (catenacyber) [inline infra/base-images/base-runner/Dockerfile]: Copied from base-builder

Maybe we can reuse the binaries already present in /out/root/go...

File: infra/base-images/base-runner/coverage
Comment (catenacyber) [inline infra/base-images/base-runner/coverage]: Maybe we can have a better way to know if we are dealing with a golang project...

File: infra/base-images/base-runner/coverage
Comment (catenacyber) [inline infra/base-images/base-runner/coverage]: We could keep the different coverage files by naming them `${fuzz_target}.cov`

File: infra/base-images/base-runner/coverage
Comment (catenacyber) [inline infra/base-images/base-runner/coverage]: If I get it right, the logic for C projects is to run the fuzz targets, and then merge coverage profiles.
We can do this for golang, it seems there is a project that does it with 111 lines

Or we can, from the start, have the command `go test -v -coverprofile` running all the fuzz targets

**Comment by Dor1s** [comment]
Thanks a lot for working on this! I'll take another look some time later

**Comment by catenacyber** [comment]
I tested the profile merging from multiple targets for go-dns project and it works :-)

I used https://github.com/wadey/gocovmerge
Is it ok ? ie no license issues ?

By the way, it seems the corpus downloads are broken
https://oss-fuzz.com/fuzzer-stats?project=gonids&fuzzer=libFuzzer_gonids_fuzz_parserule&job=libfuzzer_asan_gonids&group_by=by-day
Does not show corpus for March 9 and 10 (it does for the previous days)
And the link gets me to the error "Authorization required"

**Comment by Dor1s** [comment]
@catenacyber regarding corpus, I think corpus pruning task was somewhat broken last two days (//cc @mbarbella-chromium @mpherman2), that's why. It should recover today and the corpora links should be available again.

Regarding coverage -- really appreciate your contributions, but can't commit to a thorough review just yet.

**Comment by catenacyber** [comment]
Hi @Dor1s 
Friendly ping on this :-)
Do you know when you will have time to review it ?

**Comment by Dor1s** [comment]
@catenacyber thanks for another ping. This is not forgotten, but I can't give you any ETA, as I don't think it's worth investing into finalizing and maintaining this until we get more projects integrated (i.e. more potential users).

When Go coverage becomes more needed, we'll have to update code coverage job, ClusterFuzz side of things, etc -- this won't be done with a single PR :)

**Comment by catenacyber** [comment]
By the way (not sure where to post this), any thoughts on improving golang standard library fuzzing cf https://github.com/golang/go/issues/37499 ?
cc @inferno-chromium

File: infra/base-images/base-builder/Dockerfile
Comment (Dor1s) [inline infra/base-images/base-builder/Dockerfile:109]: do we need this scrip tin the base-builder image?

File: infra/base-images/base-runner/Dockerfile
Comment (Dor1s) [reply infra/base-images/base-runner/Dockerfile:68]: so we need golang to be installed in the base-runner image, as it's being used for generating a coverage report?

File: infra/base-images/base-runner/coverage
Comment (Dor1s) [reply infra/base-images/base-runner/coverage:118]: Yes, we now have a `language` attribute in every project.yaml file, see https://github.com/google/oss-fuzz/blob/475f86dbc4f245c3ce9b329176e2603654246d6e/infra/helper.py#L294 and how it's used

File: projects/go-dns/build.sh
Comment (Dor1s) [inline projects/go-dns/build.sh:24]: something like this needs to be done for each Go project, right?

Sadly, we decided not to put `compile_fuzzer` function anywhere globally and instead copy-pasted it into every Go project's build_file. Now is a moment to reconsider that decision

**Review by Dor1s — COMMENTED** [review]
@catenacyber good news: we decided that we should get this change landed so that other OSS-Fuzz users could generate coverage reports for Go projects locally. We don't have a concrete plan for when we'll add the infrastructure support for generating coverage reports for non-C/C++ projects, but at least for the local use case it seems worthy to do now.

May I ask you please to rebase your change on the newest tip of tree and test if it's still working?

File: infra/base-images/base-builder/Dockerfile
Comment (catenacyber) [reply infra/base-images/base-builder/Dockerfile:109]: We need this somewhere.
It is meant to be generic for all projects. (but it can be in each project)
So, it is related to `compile_fuzzer` being generic or not

What do you want for now ? specific by projet or generic ?

File: projects/go-dns/build.sh
Comment (catenacyber) [reply projects/go-dns/build.sh:24]: Yes, this needs to be done for each project.

Do you want this generalization be part of this PR ? (or rather in another one)

**Comment by catenacyber** [comment]
Great, thanks for this news

I will do the rebase and check work.

> so we need golang to be installed in the base-runner image, as it's being used for generating a coverage report?

Exactly.

> Yes, we now have a language attribute in every project.yaml file, see

I noticed but that was not the case when this work was started ;-)
I will use it in the rework

File: infra/base-images/base-builder/Dockerfile
Comment (Dor1s) [reply infra/base-images/base-builder/Dockerfile:109]: Leave it like this for now, I need to think more about how we can unify the whole compile_fuzzer thing

**Review by Dor1s — COMMENTED** [review]
> I noticed but that was not the case when this work was started ;-)

Yep, that was added after you implemented this.

**Comment by catenacyber** [comment]
> so we need golang to be installed in the base-runner image, as it's being used for generating a coverage report?

Not doing it right now, but you may want to do a gocov-runner image which will be the base-runner + golang
I think that kind of the same is done for msan-builder, right ?

**Comment by catenacyber** [comment]
Okay, So I rebased, made some fixes, and tested and it works :
First I need to run
```
docker build -t gcr.io/oss-fuzz-base/base-builder infra/base-images/base-builder
docker build -t gcr.io/oss-fuzz-base/base-runner infra/base-images/base-runner
```
Then, given the fact that I downloaded some corpus to `build/corpus/$PROJECT_NAME/$fuzz_target_name` I can run
```
export PROJECT_NAME=gonids
python infra/helper.py build_fuzzers --sanitizer=coverage $PROJECT_NAME
python infra/helper.py coverage --no-corpus-download $PROJECT_NAME
```

ie, I get the html report named as build/out/gonids/report/index.html

The output shows 
```
Running go target
=== RUN   TestFuzzParseRuleCorpus
--- PASS: TestFuzzParseRuleCorpus (15.37s)
PASS
coverage: 65.7% of statements
ok  	github.com/google/gonids	15.385s
End for go targets
```

This works also for go-dns project

I did not try to serve the report cf `python3 -m http.server $HTTP_PORT`

**Comment by catenacyber** [comment]
It should be good now.
Any thoughts about the gocovmerge path ?

File: infra/base-images/base-builder/Dockerfile
Comment (Dor1s) [reply infra/base-images/base-builder/Dockerfile:149]: Let's rename it to `ossfuzz_coverage_runner.go`

File: infra/base-images/base-runner/Dockerfile
Comment (Dor1s) [reply infra/base-images/base-runner/Dockerfile:69]: @oliverchang as we also install Go in the base-runner, would you be ok with us doing that once in the `base-image` instead?

File: infra/base-images/base-runner/coverage
Comment (Dor1s) [inline infra/base-images/base-runner/coverage:114]: if we decide to install Go in the base-image (the comment above), we should set the `GOPATH` there too. Then, here we will do `GOPATH=$OUT/$GOPATH` to keep things less fragile

File: infra/base-images/base-runner/coverage
Comment (Dor1s) [inline infra/base-images/base-runner/coverage:126]: The existing code runs fuzz targets in parallel. Can we run Go targets in parallel too?

**Review by Dor1s — COMMENTED** [review]
Awesome stuff! I've tested locally and it works!

File: infra/base-images/base-runner/coverage
Comment (catenacyber) [reply infra/base-images/base-runner/coverage]: It looks like it works :-)
Refactored the code to share the multiprocess handling

File: infra/base-images/base-runner/coverage
Comment (catenacyber) [reply infra/base-images/base-runner/coverage]: Waiting for the decision about where to put go, otherwise ok

Comment (Dor1s) [reply infra/base-images/base-runner/coverage]: Let me ping @oliverchang once again, as GitHub reviews are terrible. Oliver, we need to have Go installed in both the builder and the runner images. Would you be concerned if we install Go only once in the `base-image` instead of `base-builder` and `base-runner`?

File: infra/base-images/base-runner/Dockerfile
Comment (oliverchang) [reply infra/base-images/base-runner/Dockerfile:69]: Leaving it here is probably fine. 

We plan to diverge images here and migrate runners to be based on ClusterFuzz images instead, so we'd have to add it back here anyway.

Comment (Dor1s) [reply infra/base-images/base-runner/Dockerfile:69]: ack, thanks Oliver!

File: infra/base-images/base-builder/Dockerfile
Comment (Dor1s) [reply infra/base-images/base-builder/Dockerfile:149]: and instead of hardcoded `/root/go/` let's use `$GOPATH` which is set above on line 65. As you move this to the `base-runner` image, you'll initialize `$GOPATH` there too and then please use it instead of the hardcoded path

File: infra/base-images/base-runner/Dockerfile
Comment (Dor1s) [inline infra/base-images/base-runner/Dockerfile:27]: `base-image` does this command already. Are you sure we need it here as well?

File: infra/base-images/base-builder/Dockerfile
Comment (Dor1s) [reply infra/base-images/base-builder/Dockerfile:149]: to install this script in the `base-runner` image, add `COPY ossfuzz_coverage_runner.go /usr/local/src` on line 58 in `base-runner/Dockerfile`

File: infra/base-images/base-runner/coverage
Comment (Dor1s) [inline infra/base-images/base-runner/coverage:114]: as we install Go in `base-runner` itself, I don't think we need to re-define `GOPATH` here. Am I wrong?

Comment (Dor1s) [reply infra/base-images/base-runner/coverage:114]: if we need to, let's do `GOPATH=$OUT/$GOPATH` to avoid hardcoding

File: infra/base-images/base-runner/coverage
Comment (Dor1s) [inline infra/base-images/base-runner/coverage:162]: please add one indentation level for all the commands from here down to line 190

File: projects/go-dns/build.sh
Comment (Dor1s) [reply projects/go-dns/build.sh:23]: let's do it in another PR

**Review by Dor1s — COMMENTED** [review]
A few more comments. Should be good to land after that! I'll outline the remaining work in https://github.com/google/oss-fuzz/issues/2817

File: infra/base-images/base-builder/ossfuzz_coverage_runner.go
Comment (catenacyber) [reply infra/base-images/base-builder/ossfuzz_coverage_runner.go:1]: ok

File: infra/base-images/base-builder/Dockerfile
Comment (catenacyber) [reply infra/base-images/base-builder/Dockerfile:149]: > let's move this file to the base-runner image, as we don't really need it in the builder

We do need this file during the build process.
It is copied, then modified in every call to `compile_fuzzer`, kind of producing the coverage fuzz target

> and instead of hardcoded /root/go/ let's use $GOPATH

ok for this

File: infra/base-images/base-runner/Dockerfile
Comment (catenacyber) [reply infra/base-images/base-runner/Dockerfile:27]: I do not think so, this must have been a debug leftover

File: infra/base-images/base-runner/coverage
Comment (catenacyber) [reply infra/base-images/base-runner/coverage:162]: ok, I had not done this at first for easier review ;-)

File: infra/base-images/base-runner/coverage
Comment (catenacyber) [reply infra/base-images/base-runner/coverage:114]: > as we install Go in base-runner itself, I don't think we need to re-define GOPATH here. Am I wrong?

I am afraid you are.
There are indeed 2 GOPATHs in base-runner :
- the system one (/root/go), which is used by gocovmerge
- the fuzz targets one (/root/go/root/go), which is used by running the targets

Do you see a better way to do this ?

File: infra/base-images/base-builder/compile
Comment (catenacyber) [reply infra/base-images/base-builder/compile:95]: Maybe it can be just $GOPATH/src and $GOPATH/pkg but I am not sure...

**Comment by catenacyber** [comment]
@Dor1s I pushed some changes.
There is still the question of base-builder GOPATH copy and use in base-runner...
I left the projects changes in one commit until the question is resolved

File: infra/base-images/base-runner/coverage
Comment (Dor1s) [reply infra/base-images/base-runner/coverage:162]: smart!

File: infra/base-images/base-runner/coverage
Comment (Dor1s) [inline infra/base-images/base-runner/coverage:160]: change the log message to "Finished generating code coverage report for Go fuzz targets."

File: infra/base-images/base-runner/coverage
Comment (Dor1s) [reply infra/base-images/base-runner/coverage]: Is it important to use `/root/go/root/go` for running fuzz targets?

File: infra/base-images/base-builder/compile
Comment (Dor1s) [reply infra/base-images/base-builder/compile:95]: ok, let's leave it this way

File: infra/base-images/base-builder/Dockerfile
Comment (Dor1s) [reply infra/base-images/base-builder/Dockerfile]: > We do need this file during the build process.

Makes sense, thanks! Ok to leave here then

**Review by Dor1s — COMMENTED** [review]
One more nit and one more question left!

**Comment by catenacyber** [comment]
> Is it important to use /root/go/root/go for running fuzz targets?

Answering here as Github seems to be buggy...

It is /out/root/go in facts cf `export GOPATH=$OUT/root/go`

It is important to have both GOPATHS : this one for the targets, and the system one for gocovmerge and such

File: infra/base-images/base-runner/coverage
Comment (catenacyber) [reply infra/base-images/base-runner/coverage]: ok

**Comment by catenacyber** [comment]
> One more nit and one more question left!

Done

Should I now remove the commit "Enables golang coverage report for gonids and go-dns" ?
What about the commit "Generates summary for golang coverage reports" ?
And what about the commit "Performance profile for golang projects" ?

Do you want them in separated/later PRs ?

**Comment by catenacyber** [comment]
Retested on a new machine and it seems we need `apt-get update` in infra/base-images/base-runner/Dockerfile

File: infra/base-images/base-runner/Dockerfile
Comment (Dor1s) [reply infra/base-images/base-runner/Dockerfile:27]: thanks for verifying that it's needed!

**Review by Dor1s — APPROVED** [review]
Looks great! Testing locally now before merging.

**Comment by AdamKorcz** [comment]
This is great! I can't wait to use it.

**Pull request merged**
